webapp/src/Controller/PublicController.php

```diff
@@ -267,4 +286,129 @@ protected function getBinaryFile(int $probId, callable $response): StreamedRespo
+    {
+        $contest = $this->dj->getCurrentContest(onlyPublic: true);
+
+        if (!$contest) {
+            throw $this->createNotFoundException('No active contest found');
+        }
+
+        $scoreboard = $this->scoreboardService->getScoreboard($contest);
+
+        /** @var Submission[] $submissions */
+        $submissions = $this->submissionService->getSubmissionList(
+            [$contest->getCid() => $contest],
+            restrictions: new SubmissionRestriction(valid: true),
+            paginated: false
+        )[0];
+
+        $submissionData = [];
+
+        // We prepend IDs with team- and problem- to make sure they are not
+        // consecutive integers
+        foreach ($scoreboard->getTeamsInDescendingOrder() as $team) {
+            if ($teamId && $teamId !== $team->getExternalid()) {
+                continue;
+            }
+            $teamKey = 'team-' . $team->getExternalid();
+            $submissionData[$teamKey] = [];
+            foreach ($scoreboard->getProblems() as $problem) {
+                if ($problemId && $problemId !== $problem->getExternalId()) {
+                    continue;
+                }
+                $problemKey = 'problem-' . $problem->getExternalId();
+                $submissionData[$teamKey][$problemKey] = [];
+            }
+        }
+
+        $verificationRequired = $this->config->get('verification_required');
+
+        foreach ($submissions as $submission) {
+            $teamKey = 'team-' . $submission->getTeam()->getExternalid();
+            $problemKey = 'problem-' . $submission->getProblem()->getExternalid();
+            if ($teamId && $teamId !== $submission->getTeam()->getExternalid()) {
+                continue;
+            }
+            if ($problemId && $problemId !== $submission->getProblem()->getExternalid()) {
+                continue;
+            }
+            $submissionData[$teamKey][$problemKey][] = [
+                'time' => $this->twigExtension->printtime($submission->getSubmittime(), contest: $contest),
+                'language' => $submission->getLanguageId(),
+                'verdict' => $this->submissionVerdict($submission, $contest, $verificationRequired),
+            ];
+        }
+
+        return new JsonResponse([
+            'submissions' => $submissionData,
+        ]);
+    }
+
+    protected function submissionVerdict(
+        Submission $submission,
+        Contest $contest,
+        bool $verificationRequired
+    ): string {
+        if ($submission->getSubmittime() >= $contest->getEndtime()) {
+            return $this->twigExtension->printResult('too-late');
+        }
+        if ($contest->getFreezetime() && $submission->getSubmittime() >= $contest->getFreezetime() && !$contest->getFreezeData()->showFinal()) {
+            return $this->twigExtension->printResult('');
+        }
+        if (!$submission->getJudgings()->first() || !$submission->getJudgings()->first()->getResult()) {
```

Comment:
Cant we `!$submission->getJudgings()->first()?->getResult()` here?

Reply:
No, since `first()` returns `false|T`, not `null|T`. I had that before and it got angry 😢